File: Frontend/src/app/service/room.service.ts
import { Injectable } from '@angular/core';
import { PATH_V1 } from '../constant/Abstract.constant';
import { HttpClient } from '@angular/common/http';
import { Observable } from 'rxjs';
import { environment } from 'src/environments/environment';

@Injectable({
  providedIn: 'root',
})
export class RoomService {
  URL = environment.apiUrl + PATH_V1 + '/room';

  constructor(private httpClient: HttpClient) {}

  createNewRoom(data: any): Observable<any> {
    const formData: FormData = new FormData();
    formData.append('accommodationId', data.accommodationId);
    formData.append('roomType', data.roomType);
    formData.append('roomArea', data.roomArea);
    formData.append('bed', data.bed);
    formData.append('capacity', data.capacity);
    formData.append('smoke', data.smoke);
    formData.append('quantity', data.quantity);
    formData.append('price', data.price);
    formData.append('discount', data.discount);
    for (let i = 0; i < data.dinningRooms.length; i++) {
      formData.append('dinningRooms', data.dinningRooms[i]);
    }
    for (let i = 0; i < data.bathRooms.length; i++) {
      formData.append('bathRooms', data.bathRooms[i]);
    }
    for (let i = 0; i < data.roomServices.length; i++) {
      formData.append('roomServices', data.roomServices[i]);
    }
    for (let i = 0; i < data.views.length; i++) {
      formData.append('views', data.views[i]);
    }
    for (let i = 0; i < data.images.length; i++) {
      formData.append('files', data.images[i]);
    }

    return this.httpClient.post(this.URL + '/save', formData);
  }

  getById(roomId: number) {
    return this.httpClient.get(`${this.URL}/get-by-id/${roomId}`);
  }

  getByAccommodationId(accommodationId: number) {
    return this.httpClient.get(`${this.URL}/get-by-accommodation-id/${accommodationId}`);
  }
}

File: backend/src/main/java/vn/spring/booking/service/RoomService.java
package vn.spring.booking.service;

import com.fasterxml.jackson.core.JsonProcessingException;
import com.fasterxml.jackson.databind.ObjectMapper;
import vn.spring.booking.common.Constant;
import vn.spring.booking.constant.enums.ResponseCode;
import vn.spring.booking.utils.ObjectUtils;
import vn.spring.booking.dto.request.CreateUpdateRoomRequest;
import vn.spring.booking.dto.response.RoomResponse;
import vn.spring.booking.entities.Accommodation;
import vn.spring.booking.entities.File;
import vn.spring.booking.entities.Room;
import vn.spring.booking.repository.AccommodationRepository;
import vn.spring.booking.repository.RoomRepository;
import lombok.RequiredArgsConstructor;
import lombok.SneakyThrows;
import org.springframework.http.HttpStatus;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;
import org.springframework.util.CollectionUtils;
import vn.spring.common.constants.enums.BaseResponseCode;
import vn.spring.common.exception.BaseException;

import java.util.*;
import java.util.stream.Collectors;

@Service
@RequiredArgsConstructor
public class RoomService {
  private final RoomRepository roomRepository;
  private final AccommodationRepository accommodationRepository;
  private final FileService fileService;

  private final ObjectMapper mapper;

  @Transactional
  public void createRoom(CreateUpdateRoomRequest request) throws JsonProcessingException {
    Accommodation accommodation =
        accommodationRepository
            .findById(request.getAccommodationId())
            .orElseThrow(() -> new BaseException(BaseResponseCode.NOT_FOUND, HttpStatus.NOT_FOUND));

    List<Room> checkRooms =
        roomRepository.findByAccommodationIdAndRoomType(
            request.getAccommodationId(), request.getRoomType());
    if (!CollectionUtils.isEmpty(checkRooms)) {
      throw new BaseException(ResponseCode.BAD_REQUEST_ROOM_TYPE_EXISTED, HttpStatus.NOT_FOUND);
    }

    Room room = new Room();
    room.setRoomType(request.getRoomType());
    room.setRoomArea(Double.parseDouble(request.getRoomArea()));
    room.setBed(request.getBed());
    room.setCapacity(Integer.parseInt(request.getCapacity()));
    room.setDinningRoom(mapper.writeValueAsString(request.getDinningRooms()));
    room.setBathRoom(mapper.writeValueAsString(request.getBathRooms()));
    room.setRoomService(mapper.writeValueAsString(request.getRoomServices()));
    room.setSmoke(Boolean.parseBoolean(request.getSmoke()));
    room.setView(mapper.writeValueAsString(request.getViews()));
    room.setPrice(Double.parseDouble(request.getPrice()));
    room.setDiscountPercent(Double.parseDouble(request.getDiscount()));
    room.setQuantity(Integer.parseInt(request.getQuantity()));
    room.setStatus(Constant.STATUS_ACTIVE);
    room.setAccommodation(accommodation);
    accommodation.setRooms(new ArrayList<>(List.of(room)));
    roomRepository.save(room);
    accommodationRepository.save(accommodation);

    fileService.executeSaveImages(
        request.getFiles(),
        Constant.FILE_PREFIX_ROOM,
        String.valueOf(room.getId()),
        ObjectUtils.extractTableName(Room.class));
  }

  public RoomResponse getById(UUID id) {
    Room room =
        roomRepository
            .findById(id)
            .orElseThrow(() -> new BaseException(BaseResponseCode.NOT_FOUND, HttpStatus.NOT_FOUND));
    return transferToObject(room);
  }

  public List<RoomResponse> getByAccommodationId(Long accommodationId) {
    List<Room> rooms = roomRepository.findByAccommodationId(accommodationId);
    if (CollectionUtils.isEmpty(rooms)) {
      return new ArrayList<>();
    }
    return rooms.stream().map(this::transferToObject).collect(Collectors.toList());
  }

  @SneakyThrows
  public RoomResponse transferToObject(Room room) {
    List<File> files =
        fileService.getFilesByEntityIdAndEntityName(
            String.valueOf(room.getId()), ObjectUtils.extractTableName(Room.class));
    List<String> fileBytes =
        files.stream()
            .map(file -> fileService.encodeImageFileToString(file.getFilePath()))
            .toList();
    return RoomResponse.builder()
        .roomId(room.getId())
        .accommodationId(room.getAccommodation().getId())
        .roomType(room.getRoomType())
        .roomArea(room.getRoomArea())
        .bed(room.getBed())
        .capacity(room.getCapacity())
        .smoke(room.getSmoke())
        .price(room.getPrice())
        .discountPercent(room.getDiscountPercent())
        .quantity(room.getQuantity())
        .status(room.getStatus())
        .views(mapper.readValue(room.getView(), Set.class))
        .dinningRooms(mapper.readValue(room.getDinningRoom(), Set.class))
        .bathRooms(mapper.readValue(room.getBathRoom(), Set.class))
        .roomServices(mapper.readValue(room.getRoomService(), Set.class))
        .filePaths(fileBytes)
        .build();
  }
}
